### PR TITLE
Paul PR Clusterer

### PR DESCRIPTION
**Pull request by Paul-Hi**
- SixPlane
- CPUPlane

### CODE REVIEW

File: src/saiga/core/geometry/plane.inl [hunk 1]
```diff
@@ -28,6 +28,13 @@ inline HD Plane::Plane(const vec3& p1, const vec3& p2, const vec3& p3)
     d      = dot(p1, this->normal);
 }
 
+inline HD Plane Plane::invert() const
+{
+    vec3 pointOnPlane = getPoint();
+    Plane invPlane(pointOnPlane, -normal);
```
Comment (darglein) [inline src/saiga/core/geometry/plane.inl:34]: Why not just -normal, -d?

File: src/saiga/core/geometry/plane.inl [hunk 2]
```diff
@@ -50,4 +57,12 @@ inline HD float Plane::sphereOverlap(const vec3& c, float r) const
     return r - distance(c);
 }
 
+inline HD vec4 Plane::intersectingCircle(const vec3& c, float r) const
```
Comment (darglein) [inline src/saiga/core/geometry/plane.inl:60]: Return as std::pair<vec3,float> to make it more clear

File: src/saiga/core/geometry/sphere.cpp [hunk 3]
```diff
@@ -68,6 +68,15 @@ float Sphere::sdf(vec3 p) const
     return (p - pos).norm() - r;
 }
 
+vec2 Sphere::projectedIntervall(const vec3& d) const
+{
+    vec2 ret;
+    float t = dot(d, pos);
+    ret[0]  = std::min(t - r, t + r);
+    ret[1]  = std::max(t + r, t - r);
```
Comment (darglein) [inline src/saiga/core/geometry/sphere.cpp:76]: if (r >= 0) the min/max is not required. Better to remove min/max and add a SAIGA_ASSERT(r >=0);

File: src/saiga/opengl/rendering/lighting/light_clusterer.h [hunk 4]
```diff
@@ -285,5 +264,10 @@ class SAIGA_OPENGL_API Clusterer
     std::shared_ptr<BuildClusterComputeShader> buildClusterShader3D;
     std::shared_ptr<LightAssignmentComputeShader> buildLightToClusterMapShader;
 #endif
+    virtual bool fillImGui();
+
+   private:
+    void beginImGui(bool* p_open = NULL);
```
Comment (darglein) [inline src/saiga/opengl/rendering/lighting/light_clusterer.h:270]: why does this class need 3 imgui functions? Shouldn't one be enough?

**Review by darglein — CHANGES_REQUESTED** [review]

File: src/saiga/core/geometry/plane.inl (hunk 1 above)
Comment (Paul-Hi) [reply src/saiga/core/geometry/plane.inl:34]: You are right, change was done while debugging. Will revert that.

File: src/saiga/opengl/rendering/lighting/light_clusterer.h (hunk 4 above)
Comment (Paul-Hi) [reply src/saiga/opengl/rendering/lighting/light_clusterer.h:270]: A clusterer has some "basic" ui elements, like "TileSize", this is in the light_clusterer imgui function. But some of the derived classes have some extra ui for example "lightDebug" in the cpu_plane_clusterer. I did not want to duplicate the code in the derived classes, but with one function, I would have to start an new ImGui-Window for each derived class. So I let the base clusterer begin the window. fill it with the basic stuff, the derived stuff and close the window again. :)

**Review by darglein — APPROVED** [review]

**Pull request merged**